### PR TITLE
Barriers

### LINKED ISSUE #583: Control barriers
Upstream - https://github.com/gpuweb/gpuweb/issues/1374
Example - https://github.com/gfx-rs/wgpu/issues/1264
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: UnsupportedInstruction(Function, ControlBarrier)', bin/convert.rs:108:64

### PR DESCRIPTION
**Pull request by kvark**
Fixes #583

### CODE REVIEW

**Comment by acowley** [comment]
I just tried out this branch rather than the barrier support I've been using, and it seems to be working well! One question: why are the WGSL built-in functions in this PR named differently than the [WGSL spec](https://gpuweb.github.io/gpuweb/wgsl/#sync-builtin-functions)?

**Comment by kvark** [comment]
Oh snap, I was looking at https://github.com/gpuweb/gpuweb/pull/1449/files#diff-709b5c343d59ae10cbabe69f9df1eee9e2dbb6e9d848632fa94357083d6e0a34R5826

**Comment by kvark** [comment]
@acowley thank you for diligence! Fixed now :)

**Review by grovesNL — APPROVED** [review]
Thanks!

File: src/back/spv/writer.rs [hunk 1]
```diff
@@ -1019,6 +983,29 @@ impl Writer {
         Ok(id)
     }
 
+    fn get_index_constant(
+        &mut self,
+        index: Word,
+        types: &Arena<crate::Type>,
+    ) -> Result<Word, Error> {
+        self.get_constant_scalar(crate::ScalarValue::Uint(index as _), 4, types)
+    }
```
Comment (jimblandy) [inline src/back/spv/writer.rs:992]: This changes the types of indexes used as LOD selectors and `OpAccessChain` operands from `Sint` to `Uint`. I couldn't figure out the rules for LOD selection, but `OpAccessChain` says the indices are "treated as signed", so is this what we want?

**Review by jimblandy — COMMENTED** [review]
It looks good overall. Three concerns:
- It seems like control-only barriers will be dropped in GLSL and MSL output, which is inconsistent with the handling of SPIR-V.
- I think `Statement::Barrier` needs better docs. The way it's being used by the front and back ends seems inconsistent, so I wasn't confident that I could infer the intent from the code.
- The signedness of cached index constants in SPIR-V output changed, which is a bit surprising.

File: src/front/spv/mod.rs [hunk 2]
```diff
@@ -2046,6 +2046,42 @@ impl<I: Iterator<Item = u32>> Parser<I> {
...
+                        let mut flags = crate::Barrier::empty();
+                        flags.set(
+                            crate::Barrier::STORAGE,
+                            semantics & spirv::MemorySemantics::UNIFORM_MEMORY.bits() != 0,
+                        );
+                        flags.set(
+                            crate::Barrier::WORK_GROUP,
+                            semantics & spirv::MemorySemantics::WORKGROUP_MEMORY.bits() != 0,
+                        );
+                        block.push(crate::Statement::Barrier(flags));
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/front/spv/mod.rs:2080]: Yes, we definitely want a promotion like this

[the remark this answers is not in the document]
Comment (kvark) [reply src/front/spv/mod.rs:2080]: actually, I don't know what `SubgroupMemory` is, and SPIR-V spec isn't clear on this

File: src/back/glsl/mod.rs [hunk 3]
```diff
@@ -1398,9 +1398,15 @@ impl<'a, W: Write> Writer<'a, W> {
             // This is one of the places were glsl adds to the syntax of C in this case the discard
             // keyword which ceases all further processing in a fragment shader, it's called OpKill
             // in spir-v that's why it's called `Statement::Kill`
-            Statement::Kill => {
-                write!(self.out, "{}", INDENT.repeat(indent))?;
-                writeln!(self.out, "discard;")?
+            Statement::Kill => writeln!(self.out, "{}discard;", INDENT.repeat(indent))?,
+            // Just issue a memory barrier, ignoring the flags.
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/glsl/mod.rs:1402]: stale comment :)

File: src/lib.rs [hunk 4]
```diff
@@ -871,6 +884,8 @@ pub enum Statement {
     Return { value: Option<Handle<Expression>> },
     /// Aborts the current shader execution.
     Kill,
+    /// Synchronize accesses within the work group.
+    Barrier(Barrier),
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/lib.rs:888]: Agreed. And also, the GLSL and MSL are going to be fixed now to actually issue execution barriers.

File: src/back/spv/writer.rs (hunk 1 above)
Comment (kvark) [reply src/back/spv/writer.rs:992]: SPIR-V doesn't care, basically. It will reinterpret them as signed, and it's still fine.

**Pull request merged**